src/include/sof/audio/source_api.h

```diff
@@ -0,0 +1,180 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
+ *
+ */
+
+#ifndef __SOF_SOURCE_API_H__
+#define __SOF_SOURCE_API_H__
+
+#include <sof/common.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+/**
+ * this is a definition of API to source of audio data
+ *
+ * THE SOURCE is any component in the system that have data stored somehow and can give the
+ * data outside at request. The source API does not define who and how has produced the data
+ *
+ * The user - a module - sees this as a producer that PROVIDES data for processing
+ * The IMPLEMENTATION - audio_stream, DP Queue - sees this API as a destination it must send data to
+ * *
+ * Examples of components that should expose source API:
+ *  - DMIC
+ *	Data are coming from the outside world, stores in tmp buffer and can be presented
+ *	to the rest of the system using source_api
+ *  - a memory ring buffer
+ *	Data are coming from other module (usually using sink_api, but it does not matter in fact)
+ *
+ *  The main advantage of using source API instead of just taking pointers to the data is that
+ *  the data may be prepared at the moment the data receiver is requesting it. i.e.
+ *   - cache may be written back/invalidated if necessary
+ *   - data may be moved from circular to linear space
+ *   - part of the buffer may be locked to prevent writing
+ *   etc.etc. it depends on implementation of the data source
+ *
+ * Data in general are provided as a circular buffer and the data receiver should be able to
+ * deal with it. Of course if needed an implementation of source providing linear data can be
+ * implemented and used as a mid-layer for modules needing it.
+ *
+ * NOTE: the module should get a complete portion of data it needs for processing, process it
+ *       than release. The reason is - the depending on the implementation, the calls may be
+ *       expensive - may involve some data moving in memory, cache writebacks, etc.
+ *
+ * Multicore:
+ *  single source is intended to be used by a single component only, so all calls to it
+ *  may safely be performed using a cached pointer alias. All data modified by source helper are
+ *  local and specific to a single source only
```

Comment:
this is very very unclear. What does single source mean really? as compared to 'multi sources'?
And no word on the 'locked attribute' ?